feat: exchange cloud ID

redblom opened the pull request:
**~~This is a work in progress~~**
This PR implements https://github.com/nextcloud/contacts/issues/4416

### About
The goal of this feature is to allow for easy remote contact information exchange (cloud ID, email, name) between 2 users from different instances/organizations based on an invitation request.
The feature is an implementation of the invitation workflow from the OCM specification (https://datatracker.ietf.org/doc/draft-ietf-ocm-open-cloud-mesh/)<br><br>

A user (sender) creates the invitation. This invitation consists of an invite link and (encoded) invite code. The invitation is sent to another user (receiver) via email or through another channel, eg. chat.
After receiving the invitation, the receiver can accept the invitation by following the invite link or by pasting the (encoded) invite code into a manual accept form. 
After accepting, the exchange of user information will take place (OCM `/invite-accepted` endpoint) resulting in the creation of both users in each other's contacts list.
Following the invite link in the email is the most convenient way to accept. It will redirect the receiver via a WAYF (Where Are You From) page immediately to an accept dialog.  
Note: Manual acceptance of the invitation works with both the invite link and (encoded) invite code.

### Screenshots
<img width="1865" height="905" alt="1_contacts-app-main-page-Mike" src="https://github.com/user-attachments/assets/f067d3ef-aea0-453a-8c63-c07276071de3" />
Contacts page with new buttons 'Invite contact', 'Accept invite' and 'All invites'.<br><br>

<img width="1864" height="905" alt="2_fill-in-invite-form" src="https://github.com/user-attachments/assets/3d53aa27-0e6e-4416-ad03-709062d92115" />
Click 'Invite contact' and fill in form fields:

- Invite label for reference in the Invites list (optional)
- Recipient email (required)
- Personal message that will be sent with the email (optional)
- Check to send copy of the invite email to yourself (optional)

<img width="1865" height="905" alt="3_invite-sent-list" src="https://github.com/user-attachments/assets/3c267284-2063-4235-bf46-1264264fec95" />
'All invites' button displays the invites sent.<br><br>

<img width="1865" height="905" alt="3a_alt-way-to-share-invite" src="https://github.com/user-attachments/assets/21f89154-f84a-4a1d-bf1a-c26022612aab" />
It's also possible to share the invite by sending the invite link or code through an alternative channel (eg. chat).<br><br>

<img width="1113" height="818" alt="4_invite-email-to-Pete" src="https://github.com/user-attachments/assets/2fb81284-bf6f-451a-8074-b5aedaf40523" />
This what the email to the receiver of the invite looks like.<br><br>

<img width="1113" height="818" alt="5_copy-of-invite-email-to-Pete" src="https://github.com/user-attachments/assets/175e4e53-c7c7-48aa-90f1-a4eb07eff646" />
This is what the copy of the email to the sender of the invite looks like.<br><br>

<img width="1864" height="905" alt="6_wayf-page" src="https://github.com/user-attachments/assets/aa095b06-5ca3-4a15-877c-09a97e68ac69" />
Clicking on the invite link will direct the receiver to a WAYF (Where Are You From) page.<br>
Here you choose your institute. You can also fill in your institute's provider address if you are not on the list. A discovery process will run to check whether your provider supports OCM.<br><br>

<img width="1864" height="905" alt="7_login-after-wayf-page-select-Pete" src="https://github.com/user-attachments/assets/13b4ecec-2521-4fed-85b7-54fb07578d4b" />
Login after selecting your institute.<br><br>

<img width="1864" height="905" alt="8_accept-invite" src="https://github.com/user-attachments/assets/30a5129c-ef8a-4be4-a226-1a826726abc6" />
You will be redirected to an invite accept dialog.<br><br>

<img width="1864" height="905" alt="9_invite-accepted-Pete" src="https://github.com/user-attachments/assets/8a3c4f63-e1d4-4b09-be76-0d4ccc13b133" />
On acceptance the contact information exchange will take place via the OCM `/invite-accepted` endpoint. This will lead to the creation of a new contact (the sender) on the receiver's end.<br><br>

<img width="1864" height="905" alt="10_invite-accepted-Mike" src="https://github.com/user-attachments/assets/9084bbeb-d30e-4e1e-b554-79a6b5438259" />
And on the sender end the receiver contact info will also be added.<br><br>

<img width="1864" height="905" alt="11_alt-invite-without-email-Jimmie" src="https://github.com/user-attachments/assets/681e62e4-0604-44d0-ac73-447fb4beba2a" />
When the config setting `ocm_invites_optional_mail` is set to false, creating an invite will give you by default the option to share the invite through an alternative channel (eg. chat).<br><br>

<img width="1864" height="905" alt="12_alt-invite-no-email-copy-code" src="https://github.com/user-attachments/assets/31756f36-af11-414c-85de-df9e677e5661" />
Copy the invite link or code and send it via your channel of choice.<br><br>

<img width="1861" height="901" alt="12a_alt-invite-with-email" src="https://github.com/user-attachments/assets/900925d7-ccae-4d94-b92e-d0bc9b53cfaa" />
You may still decide to send the invite via email by clicking the 'Send via email' button.<br><br>

<img width="1861" height="892" alt="13_alt-manual-accept-invite-Dan" src="https://github.com/user-attachments/assets/8fb10a54-6271-4285-a989-147e52c6cfd8" />
The receiver of the invite can click 'Accept invite' and manually paste the invite link, code, or encoded invite code into the accept form.<br>
And this will lead to the exchange of contact information via the OCM /invite-accepted endpoint.<br><br>

<img width="1858" height="905" alt="14_alt-manual-invite-accepted-Dan" src="https://github.com/user-attachments/assets/0dffa2f3-e9a2-42f5-8b91-b97f46d42003" />
After contact info exchange the sender will be added as a new contact on the receiver's end.<br><br>

<img width="1868" height="906" alt="15_alt-manual-invite-accepted-Jimmie" src="https://github.com/user-attachments/assets/b20775bb-96b8-4746-aebf-df5d3b4d0b22" />
And the receiver will be added as a new contact on the sender's end.<br><br>

### OCM invites admin settings
<img width="1296" height="365" alt="ocm-invites-admin-settings" src="https://github.com/user-attachments/assets/a2065b64-60b5-46a1-a9e4-b387d20e3e44" />

redblom commented:
Current questions:
 - after pressing send button on the invitation form the `ContactDetails.editMode` property should be set to false to display the contact (same behavior as pressing the contact save button). How do we do that?
 Would a custom `PropertyCloudId` component to be used instead of `ContactDetailsProperty` for the cloud ID field make sense? 
 Or should we simply forward/bind the `editMode` flag to the `ContactDetailsProperty` --> `PropertyText` --> `CloudIdExchangeInvite` component chain.
 -

redblom commented:
> Thanks for the contribution, @redblom!
> 
> > The features relies on an implementation of the Invitation Workflow (see https://github.com/sara-nl/nc-collaboration).
> 
> Is there any way this could work without introducing an app dependency?

Short answer: No, that is to say, the actual cloud ID exchange is not a responsibility of the contacts app but the contacts app (the cloud ID field) is a logical place where to have this option (as per earlier discussion with the Design Team).
The feature actually depends on the implementation of OCM `/invite-accepted`. This is currently in the process of being implemented for nextcloud server core (https://github.com/nextcloud/server/pull/51113) so cloud ID exchange can become an integral part of nextcloud. Up until then there should be an app present implementing this.
The idea is to make the option not visible to the user when the platform can not do the actual cloud ID exchange (by configuration).

ChristophWurst commented:
Thanks for the clarification

mickenordin commented:
Adding some screenshots here to show you what it looks like, this is the dialog for adding a new invitation:
<img width="1337" height="912" alt="2025-09-25T16:01:24,021984991+02:00" src="https://github.com/user-attachments/assets/bfe0d203-5192-4f91-9c1a-3c990c9c7338" />
This is what it looks like after it is sent:
<img width="1259" height="435" alt="2025-09-25T16:01:40,518818898+02:00" src="https://github.com/user-attachments/assets/b59424a1-2907-43c2-b191-a8b6c91a61a9" />
This is the default Where-Are-You-From page, when no federation has been configured
<img width="1278" height="672" alt="2025-09-25T16:05:10,821218256+02:00" src="https://github.com/user-attachments/assets/d2694116-4db2-467c-9517-8977f8aef89b" />
This is what the WAYF-page looks like with a federation configured:
<img width="1072" height="845" alt="2025-09-25T16:06:40,570032364+02:00" src="https://github.com/user-attachments/assets/033e4bc9-e493-4152-b149-b8e48b9203fe" />
It is also possible to add multiple federations:
<img width="1088" height="864" alt="2025-09-25T16:14:34,777039486+02:00" src="https://github.com/user-attachments/assets/175973b8-d13d-4f16-92f2-594f347dc382" />
This is what the type-to-search feature looks like:
<img width="1104" height="622" alt="2025-09-25T16:14:50,748780755+02:00" src="https://github.com/user-attachments/assets/3588ce9b-4ecf-4667-aff3-311e9e9c0097" />

ArtificialOwl commented on lib/WayfProvider.php:
```diff
@@ -0,0 +1,120 @@
+
+	public function getMeshProvidersFromCache(): array {
+		$json = $this->appConfig->getValueString(Application::APP_ID, 'federations_cache');
+		$data = json_decode($json, true);
```
you can use getValueArray() / setValueArray() and set the config value as `lazy`

ArtificialOwl commented on lib/Service/FederatedInvitesService.php:
```diff
@@ -0,0 +1,57 @@
+
+	public function isOcmInvitesEnabled():bool {
+		$val = $this->appConfig->getValueBool(Application::APP_ID, 'ocm_invites_enabled', FederatedInvitesService::OCM_INVITES_ENABLED_BY_DEFAULT);
+		$boolval = (is_string($val) ? filter_var($val, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE) : (bool)$val);
```
`getValueBool()` will always returns `boolean`

ArtificialOwl commented on lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * @template-extends QBMapper<FederatedInvite>
+ */
+class FederatedInviteMapper extends DbFederatedInviteMapper {
```
not a huge fan of this,
if the DB is created and managed from cloud_federation_api, could this request be also integrated to the app, and calling an API from the cloud_federation_api to select/update/delete data from/to database ?

ArtificialOwl reviewed (commented):
quick review, not fully complete, but already some elements to rework

redblom commented:
@miaulalala - What's wrong with the sign-off, maybe co-author is not allowed ?

miaulalala commented:
> @miaulalala - What's wrong with the sign-off, maybe co-author is not allowed ?

No that's not the issue, one setting has Anton P., the other Anton P (no dot) 😅

redblom commented:
> > @miaulalala - What's wrong with the sign-off, maybe co-author is not allowed ?
> 
> No that's not the issue, one setting has Anton P., the other Anton P (no dot) 😅

Pfff :flushed:

artonge commented on appinfo/routes.php:
```diff
@@ -5,10 +5,22 @@
 		['name' => 'contacts#direct', 'url' => '/direct/contact/{contact}', 'verb' => 'GET'],
 		['name' => 'contacts#directcircle', 'url' => '/direct/circle/{singleId}', 'verb' => 'GET'],
+
+		['name' => 'federated_invites#get_invites', 'url' => '/ocm/invitations', 'verb' => 'GET'],
```
Those can be replaced by the following attribute. See suggestion in `FederatedInvitesController`

```php
	#[Route(type: Route::TYPE_FRONTPAGE, verb: 'GET', url: '/ocm/invitations')]
```

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,485 @@
+	 * @return JSONResponse
+	 */
+	#[NoAdminRequired]
+	public function getInvites(): JSONResponse {
```
```suggestion
	#[Route(type: Route::TYPE_FRONTPAGE, verb: 'GET', url: '/ocm/invitations')]
	public function getInvites(): JSONResponse {
```

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,485 @@
+	 * @return JSONResponse with data signature ['token' | 'message'] - the token of the deleted invitation or an error message in case of error
+	 */
+	#[NoAdminRequired]
+	public function deleteInvite(string $token): JSONResponse {
```
And so on

```suggestion
	#[Route(type: Route::TYPE_FRONTPAGE, verb: 'DELETE', url: '/ocm/invitations/{token}')]
	public function deleteInvite(string $token): JSONResponse {
```

artonge commented on lib/Command/DisableOcmInvites.php:
```diff
@@ -0,0 +1,41 @@
+			return self::SUCCESS;
+		}
+
+		$this->appConfig->setValueBool('contacts', 'ocm_invites_enabled', false);
```
Let's introduce a lexicon in Contacts so that new settings are tracked.

Documentation: https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/config/lexicon.html
Example: https://github.com/nextcloud-gmbh/server/blob/04172b855de465bc661034b32724ca394ffee3fe/apps/settings/lib/ConfigLexicon.php#L22

```suggestion
		$this->appConfig->setValueBool(App::ID, 'ocm_invites_enabled', false);
```

artonge commented on lib/Command/DisableOcmInvites.php:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		$this->appConfig->setValueBool('contacts', 'ocm_invites_enabled', false);
+		$this->appConfig->deleteKey('core', ConfigLexicon::OCM_INVITE_ACCEPT_DIALOG);
```
This is a clean up of the old key? If so this should go into a repair step: https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/repair.html

artonge commented on lib/Command/EnableOcmInvites.php:
```diff
@@ -0,0 +1,42 @@
+
+	protected function configure(): void {
+		$this
+			->setName('contacts:enable-ocm-invites')
```
Overall, this is nice for discoverability to have those commands, but they could be replaced by `occ config:app:set contacts ocm_invites_enabled --value false/true --type boolean`. I would consider removing them, or making it a single command like the `occ maintenance:mode` command.

artonge commented on lib/Cron/UpdateOcmProviders.php:
```diff
@@ -0,0 +1,36 @@
+
+class UpdateOcmProviders extends TimedJob {
+	// Run every five minutes
+	private int $expire_time = 5 * 60;
```
On which occasion will the data change? I am wondering if once a day would be enough. Or even let the admin trigger a command for that?

artonge replied:
Or could this be done during the Wayf request?

redblom replied:
> Or could this be done during the Wayf request?

The idea was to not do that. It's the reason to have the cache.

Further, an admin trigger would require the admin to be notified, meaning no guarantees there.
But yes we're looking for some sane default value ...

artonge replied:
> The idea was to not do that. It's the reason to have the cache.

Is this because the operation is slow?
You can also [manually register](https://docs.nextcloud.com/server/latest/developer_manual/basics/backgroundjobs.html#registering-manually) a `QueuedJob` for this kind of use case.

redblom replied:
@artonge - In the docs I read that QueuedJobs are for one time jobs that may be triggered (added/removed) following an event. In this case we are dealing with job that deals with retrieving data (the wayf providers list) from an external service. That service does not emit an event when the list of providers is updated, so we must consult it periodically. We also can't rely on the responsiveness of that service, this is why we cache the retrieved data (the list of providers).
How can QueuedJob help in this usecase?

miaulalala replied:
TimedJob is the right choice - but split it in two:

Every 15 minutes, check the list if new hosts / instances were added. Only add those to the list and check their capabilities, don't read every host.

Every 24 hours run a full sync on every other host that's in the mesh and re- write the capabilities to the cache. Outside of working hours would be best.

redblom replied:
:+1: https://github.com/nextcloud/contacts/pull/4417/commits/9142baf7210868a4a3e6116d7cf9683859fddd8a

artonge commented on lib/Migration/Version8004Date20260130131217.php:
Is the schema different from the old table? If not, could be we simply reuse the old table instead of dropping it?

redblom replied:
Addressed in https://github.com/nextcloud/server/pull/57853, see https://github.com/nextcloud/server/pull/57853#discussion_r3173052270

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,485 @@
+	 */
+	#[NoAdminRequired]
+	#[NoCSRFRequired]
+	public function inviteAcceptDialog(string $token = '', string $providerDomain = ''): TemplateResponse {
+		$this->initialState->provideInitialState('inviteToken', $token);
+		$this->initialState->provideInitialState('inviteProvider', $providerDomain);
+		$this->initialState->provideInitialState('acceptInviteDialogUrl', FederatedInvitesService::OCM_INVITE_ACCEPT_DIALOG_ROUTE);
+
+		return $this->index();
+	}
```
I would not extend the `PageController` just for that. Maybe simply redirect to the index, and add some query parameters to the URL?

redblom replied:
@artonge : you mean like so?

```
class PageController extends Controller {
	...

	public function index(string $token = '', string $providerDomain = ''): TemplateResponse {
		if(isset($token) && isset($providerDomain)) {
			return $this->handleInviteAcceptDialog($token, $providerDomain);
		}
		...
	}

	private function handleInviteAcceptDialog(string $token = '', string $providerDomain = ''): TemplateResponse {
		$this->initialState->provideInitialState('inviteToken', $token);
		$this->initialState->provideInitialState('inviteProvider', $providerDomain);
		$this->initialState->provideInitialState('acceptInviteDialogUrl', FederatedInvitesService::OCM_INVITE_ACCEPT_DIALOG_ROUTE);
		return new TemplateResponse(Application::APP_ID, 'main');
	}
...
```

artonge replied:
Yes, and looking at the variable that you provide:
- `inviteToken` seems to be accessible from the frontend already
- same for `inviteProvider`

So maybe you only need to provide `acceptInviteDialogUrl`?

redblom replied:
:+1:  https://github.com/nextcloud/contacts/pull/4417/commits/ea98855205bad78153a91291b5ef83fe5b8e4045

`inviteToken` and `inviteProvider` are not already accessible.

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,485 @@
+
+	/**
+	 * Resets the creation and expiration dates, and sends a new invite to the recipient.
+	 *
```
```suggestion
```

redblom replied:
:+1:  https://github.com/nextcloud/contacts/pull/4417/commits/eb7e7af8a4c7018f7725d5f3e77c4166eb5c53bf

artonge commented on src/components/Ocm/OcmAcceptForm.vue:
```diff
@@ -0,0 +1,129 @@
+    parseInvite(str) {
+      function tryParse(s) {
+        const idx = s.lastIndexOf("@");
+        if (idx === -1) return null;
```
Please do not save a few characters at the expense of readability.

```suggestion
        if (idx === -1) {
            return null;
        }
```

artonge replied:
Same applies to the rest of the code.

artonge commented on src/css/wayf.scss:
Can't we have that in `Wayf.vue`?

redblom replied:
:+1: https://github.com/nextcloud/contacts/pull/4417/commits/622904d00a6b85e51a933e718dd79a5160c00a7c

artonge commented on src/store/ocminvites.js:
```diff
@@ -0,0 +1,123 @@
+
+const actions = {
+    fetchOcmInvites(context) {
+		axios.get(generateUrl('/apps/contacts/ocm/invitations'))
+			.then(response => {
+				context.commit('appendInvites', response.data)
+				context.commit('sortInvites', response.data)
+			})
+			.catch((error) => {
+				logger.error('Error fetching OCM invites: ' + error)
+			})
```
We usually try to use `try/catch` nowadays

artonge commented on src/models/ocminvite.ts:
```diff
@@ -0,0 +1,59 @@
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+export default class OcmInvite {
```
Any reason to wrap the data in a class?

artonge commented on src/store/ocminvites.js:
Let's use typescript as much as possible.

artonge commented on src/wayf.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+
+document.addEventListener("DOMContentLoaded", () => {
```
I don't think you need to wrap that code in a listener.

artonge commented on src/wayf.js:
```diff
@@ -0,0 +1,23 @@
+  app.config.globalProperties.n = n;
+  app.mount("#contacts-wayf");
+});
+import './css/wayf.scss'
```
Let's move imports at the top of the file.

artonge commented on src/wayf.js:
```diff
@@ -0,0 +1,23 @@
+import { translate as t, translatePlural as n } from "@nextcloud/l10n";
+import Wayf from "./components/Ocm/Wayf.vue";
+
+if (!document.body.id) {
+  document.body.id = 'body-public';
+}
```
Why is this needed?

artonge commented on src/models/constants.ts:
```diff
@@ -130,13 +135,13 @@ export const CIRCLES_MEMBER_GROUPING = [
-		share: ShareType.Remote,
-		type: MEMBER_TYPE_USER,
-	},
+	// TODO: implement federated
+	// {
+	// id: `picker-${ShareType.Remote}`,
+	// label: t('contacts', 'federated users'),
+	// share: ShareType.Remote,
+	// type: MEMBER_TYPE_USER
+	// },
```
This is not working anymore?

artonge commented on lib/Listener/OcmDiscoveryListener.php:
```diff
@@ -0,0 +1,43 @@
+	public function handle(Event $event): void {
+		if ($event instanceof LocalOCMDiscoveryEvent) {
+			$event->addCapability('invite-accepted');
+			$inviteAcceptDialog = $this->appConfig->getValueString('core', ConfigLexicon::OCM_INVITE_ACCEPT_DIALOG);
```
Can't this be a Contacts app settings instead of consuming the private one?

redblom replied:
Addressed in the latest push.

artonge commented on lib/Service/FederatedInvitesService.php:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	public function isOcmInvitesEnabled():bool {
+		return $this->appConfig->getValueBool(Application::APP_ID, 'ocm_invites_enabled', FederatedInvitesService::OCM_INVITES_ENABLED_BY_DEFAULT);
```
Let's use a lexicon.

artonge commented on lib/Service/SocialApiService.php:
```diff
@@ -225,6 +229,64 @@ public function updateContact(string $addressbookId, string $contactId, ?string
+			}
+
+			/** @var \OCP\IAddressBook */
+			$addressBook = null;
+			$addressBooks = $this->manager->getUserAddressBooks();
+			foreach ($addressBooks as $_addressBook) {
+				// TODO properly resolve the correct addressbook to add the contact to
+				// Resolve by uri seems a bit risky ... can we be sure the uri equals 'contacts' ?
+				// Perhaps add to the first 'non system' addressbook we find ?
+				// (although we still would like to add to the 'Contacts' addressbook I guess)
+				if ($_addressBook->getUri() === 'contacts') {
+					$addressBook = $_addressBook;
+					break;
+				}
+			}
```
Maybe @ChristophWurst can give us the better way of doing this

redblom replied:
Addressed in the latest push.

artonge commented on lib/WayfProvider.php:
```diff
@@ -0,0 +1,125 @@
+	 * @return array an array containing all mesh providers
+	 */
+	public function getMeshProviders(): array {
+		$urls = preg_split('/\s+/', trim($this->appConfig->getValueString(Application::APP_ID, 'mesh_providers_service')));
```
Lexicon

artonge commented on lib/WayfProvider.php:
```diff
@@ -0,0 +1,125 @@
+	 * @return array an array containing all mesh providers
+	 */
+	public function getMeshProvidersFromCache(): array {
+		$data = $this->appConfig->getValueArray(Application::APP_ID, 'federations_cache', [], true);
```
Lexicon

miaulalala commented:
@redblom - can you check on artonge's review comments please?

redblom commented:
> @redblom - can you check on artonge's review comments please?

:+1:  working on it.

redblom commented:
> @redblom - can you check on artonge's review comments please?

@miaulalala We've now covered them all we think.

redblom commented:
Rebased and some cleanup.

miaulalala commented:
Hi @redblom! CI is currently failing on two fronts:

**1. Missing `#[Override]` attributes (Psalm)**
Multiple methods in `Cron/UpdateOcmProviders.php`, the two new Listener classes, and the Migration files are missing `#[Override]` attributes. You can fix most of these automatically by running `composer psalm:update-baseline` and committing the result.

**2. NPM lint error in `src/wayf.js`**
Line 13 uses a top-level arrow function — change it to a regular `function` declaration to satisfy the `antfu/top-level-function` rule, then run `npm run lint:fix`.

artonge commented on lib/Command/OcmInvitesConfig.php:
Do we really want a command for that? Why is the already existing command to set config values not enough?

If the reason is for documentation, then I think we should drop the custom command and simply extend the documentation.

redblom replied:
@artonge - Hmm maybe we misunderstood your suggestion here. A dedicated command (with documentation) seemed useful to us.

But is extending the docs the preferred way then?

artonge replied:
> But is extending the docs the preferred way then?

I would say yes, as in this case this is doing nothing else than setting the config value.

redblom replied:
:+1: https://github.com/nextcloud/contacts/commit/d1e34e7b382cada3bdf852df8f29b7b7e7677d22

artonge commented:
@redblom there are still some unresolved conversations from the last time I reviewed, can you address them?

redblom commented:
@artonge 
> Is this because the operation is slow?

Yes, it can be slow-ish, so definitely for a responsive wayf page.

> You can also [manually register](https://docs.nextcloud.com/server/latest/developer_manual/basics/backgroundjobs.html#registering-manually) a `QueuedJob` for this kind of use case.

Will have a look at that.

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,1066 @@
+	 * On success the user is redirected to the new contact url.
+	 *
+	 * @param string $token the token of the invite
+	 * @param string $provider the provider of the sender of the invite
```
The argument is not part of the method signature

redblom replied:
:+1:  https://github.com/nextcloud/contacts/pull/4417/commits/ea98855205bad78153a91291b5ef83fe5b8e4045

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,1066 @@
+	 * @param string $recipientAddress the original recipient's email address
+	 * @param string $message the optional message included in the invite
+	 */
+	private function sendCcEmail(string $token, string $senderProvider, string $senderAddress, string $recipientAddress, string $message): void {
```
Isn't the logic really close to the `sendMail` method?
Can't we add a `$cc` argument to `sendMail` to benefit from the work that was already done there?

redblom replied:
Indeed!
:+1: https://github.com/nextcloud/contacts/pull/4417/commits/1438c7270c3ea15ec212da3cf0b14fa076f3947f

artonge commented on lib/WayfProvider.php:
```diff
@@ -0,0 +1,148 @@
+			try {
+				$res = $this->httpClient->newClient()->get($url);
+				$code = $res->getStatusCode();
+				if (!($code >= 200 && $code < 400)) {
```
```suggestion
				if (!($code >= Http::STATUS_OK && $code < ...)) {
```

redblom replied:
:+1: https://github.com/nextcloud/contacts/pull/4417/commits/2a3d1fee57a3b70740647f52de7173fc382548a6

artonge commented on lib/WayfProvider.php:
```diff
@@ -0,0 +1,148 @@
+				if (!($code >= 200 && $code < 400)) {
+					continue;
+				}
+				$data = json_decode($res->getBody(), true);
```
Where is the format of the answer defined?

redblom replied:
:+1: https://github.com/nextcloud/contacts/pull/4417/commits/88ae1b584d3dee6de0d5d60dc338e855a7af1b41

artonge commented on lib/Controller/FederatedInvitesController.php:
```diff
@@ -0,0 +1,1066 @@
+	/**
+	 * Returns all open (not yet accepted) invites.
+	 *
+	 * @return JSONResponse
```
Phpdoc are not necessary if the native php types are used, you can remove them.

redblom replied:
?? I don't see that (much) in other places 
Also: https://docs.nextcloud.com/server/stable/developer_manual/getting_started/coding_standards/php.html#comments

nimishavijay commented:
This seems very comprehensive, nice! :) From what I understand this is a way to exchange contacts with people from other organizations directly from the contacts app. Design wise I have a few questions and some feedback:

Thinking about the design and flow itself, some questions come to mind:
- [ ] There seem to be a lot of ways of sending invitations: via email, via invite link, via invite code. Is there a necessity to have an invite code? Could we have just one way: invite link and have options like "copy link" and "share via email" for that one link? It would simplify a lot of decision making for the user. They would only have to click on a link to exchange contacts in every case. Does that make sense?
- [ ] Since the goal is to NOT look at the invites page (there shouldn't be invitations that are sent and pending acceptance, and once you have exchanged contacts there's no reason to look at the invitation again) I am not sure if having an entire view in the main content is the way to go. What do you think about moving it into a modal?
- [ ] Similar reasoning for the buttons too: the combined visual weight of the invitations is even more than of the CTA of the whole app! So what do you think about collapsing all invites into one item called "External invites" and that opens up a modal with the list of invitations and an option to create?
- [ ] instead of having an option to send a copy of the invite to the sender's own address, why not have the invite message shown in the UI? It can be collapsed by default to save space of course

Visual design feedback to be more consistent with Nextcloud design:
- [ ] We want to be accessible, so the headings for the modals/dialogs must be h2. I see that some of them are different like h5. If you want them to look smaller you can override the styles, but it should follow the heading order
- [ ] The buttons in a dialog a usually positioned as: primary action button on the right end and the cancel button on the left end for left-to-right languages. Your buttons seem to be the other way around, which could lead to accidentally clicking on cancel. The primary action button should also always be the primary variant so that the color stands out in the dialog
- [ ] Also try to use existing variables or multiples of existing variables for padding, margin and gaps as much as possible. Instead of 4px, use `var(--default-grid-baseline)` 
- [ ] In the "where are you from" page, consider using the [NcFormbox](https://nextcloud-vue-components.netlify.app/#/Components/NcFormBox) component, it would automatically adjust colours, spacing, etc


But either way this is really cool. Interested to know what you think of the design feedback! :)

redblom commented:
> But either way this is really cool. Interested to know what you think of the design feedback! :)

@nimishavijay - Thank you for this valuable feedback! We're on it.

redblom commented:
@artonge - addressed all your reviews.
I have two questions left where I would like your input:
https://github.com/nextcloud/contacts/pull/4417#discussion_r3274687716
https://github.com/nextcloud/contacts/pull/4417#discussion_r3302996398

redblom commented:
> This seems very comprehensive, nice! :) From what I understand this is a way to exchange contacts with people from other organizations directly from the contacts app. Design wise I have a few questions and some feedback:
> 
> ...

#### Re-design of the layout:
<img width="1859" height="908" alt="main-01-main drawio" src="https://github.com/user-attachments/assets/781b60bf-93a3-4d59-a563-61850382487c" />

_3 buttons collapsed into 1 'OCM invitations' and 1 permanent sub item 'All invitations' that shows the nr. of pending invitations_


<img width="1859" height="907" alt="main-01-main-dropdown drawio" src="https://github.com/user-attachments/assets/88416c4f-0312-4eba-99b8-8d52205bf7fd" />

_Sub menu displays the options to create and accept invites_

<img width="1859" height="908" alt="main-01-main-create-invitation-show-email-preview-button drawio" src="https://github.com/user-attachments/assets/aaf9f8de-91b8-47d3-a1a2-c269bcff2816" />

_A preview of the email can be shown_

<img width="1859" height="907" alt="main-01-main-create-invitation-show-email-preview drawio" src="https://github.com/user-attachments/assets/ecf52350-bde9-4fff-a861-9b5e5d191f49" />

_The email preview_

<img width="1859" height="907" alt="main-01-main-all-invitations drawio" src="https://github.com/user-attachments/assets/a2b5ad7c-2f96-42df-805c-9be38b6439eb" />

_Options to copy link, (encoded) invite_

miaulalala commented on lib/Cron/UpdateDeltaOcmProviders.php:
```diff
@@ -0,0 +1,36 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
```

Please also add a short description about what this job does

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Cron/UpdateOcmProviders.php:
```diff
@@ -0,0 +1,38 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
```

Same here, please add a description

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Db/FederatedInvite.php:
```diff
@@ -0,0 +1,78 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
```

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,183 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```
same

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Exception/ContactExistsException.php:
```diff
@@ -0,0 +1,15 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```
same

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Service/FederatedInvitesService.php:
```diff
@@ -0,0 +1,257 @@
+		} catch (DoesNotExistException) {
+			$response = ['message' => 'Invalid or non existing token', 'error' => true];
+			$status = Http::STATUS_BAD_REQUEST;
+			$response = new JSONResponse($response, $status);
```
```suggestion
			$response = new JSONResponse($response, Http::STATUS_BAD_REQUEST);
```

Same for all others

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Exception/ContactExistsException.php:
```diff
@@ -0,0 +1,15 @@
+
+use Exception;
+
+class ContactExistsException extends Exception {
```
```suggestion
class ContactAlreadyExistsException extends Exception {
```

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

miaulalala commented on lib/Service/SocialApiService.php:
```diff
@@ -234,6 +239,130 @@ public function updateContact(string $addressbookId, string $contactId, ?string
+			);
+			return $newContact;
+		} catch (Exception $e) {
+			$this->logger->error('An exception occurred creating a federated contact: ' . $e->getTraceAsString(), ['app' => Application::APP_ID]);
```
```suggestion
			$this->logger->error('An exception occurred creating a federated contact: ' . $e->getMessage()(), ['exception' => $e]);
```

redblom replied:
https://github.com/nextcloud/contacts/pull/4417/commits/06636c830032e77eceac9f0543dfe7b6341b96ea

redblom commented:
#### GUI after modifications:

<img width="1862" height="902" alt="Screenshot from 2026-06-21 15-08-59" src="https://github.com/user-attachments/assets/d032d169-d73b-4147-9b15-3fea7787f2f3" />
New External invitations header and context menu

<img width="1862" height="902" alt="Screenshot from 2026-06-21 14-34-23" src="https://github.com/user-attachments/assets/50c557e1-f504-4e64-8a51-fc00b2a1baee" />
Email preview option

<img width="1862" height="902" alt="Screenshot from 2026-06-21 14-34-37" src="https://github.com/user-attachments/assets/1486086b-5fe5-4386-8b61-6d41c28e76d3" />
Preview email

miaulalala commented:
Feedback by @jancborchardt:

**Must haves**
- [x] Make "All invitations" a list item below the address book list instead of a separate menu section:
<img width="645" height="211" alt="image" src="https://github.com/user-attachments/assets/106da0cb-0511-45d5-bdfb-03a4bd3ba1df" />

- [x] Make adding a new invitation or accepting an invitation as a separate button next to the "new contact" button - example from Talk:
<img width="396" height="285" alt="image" src="https://github.com/user-attachments/assets/7dd7e96f-3a2d-4a0e-a6b0-d04a3527a619" />

- [x] Make "email address" the first input field in the Invitation modal and remove the label
- [x] Replace "Provider" with "Select your server" as a heading. Remove subtext from heading.
- [x] Put accept buttons on the rightmost side, and be a primary button (if not fixed yet)
- [x] Add "create invitation" and "accept invitation" on empty content page when no invitations are open
- [x] Add title to Invitation and Accept modal
- [x] Remove icon from cancel button in Invitation and Accept Modal
- [x] No full stop in headings
- [ ] Give good user feedback for errors - such as an invite not being able to be accepted
	- [ ] Also hide the button to accept an invite if the capability is not there (may already be done) 

**Nice to have** - will not block merge
- [x] Remove extra container around recipient email and personal message so all form element are on one level
- [ ] Change input field for adding a custom provider: show a "+ Add custom provider" button -> show input field only when button is clicked
- [ ] Rename "Continue" button to "Accept invite" button to "provider" page
- [ ] Rename "Cancel" to "Decline invite" for Accept modal
- [ ] Reduce width of provider list on invitation acceptance page
- [ ] Remove cloud icon from provider / server list
- [ ] Add border radius consistently for "provider" page

**Future enhancement and ideas**
-  [ ] store chosen server in localstorage
-  [ ] moving federated couldId field from ...more to the immediate menu
-  hand down who is sending the invitation via link hash or similar - to be discussed with the OCM working group

jancborchardt commented:
https://github.com/nextcloud/contacts/pull/4417#issuecomment-4789621159

FYI for reference @nimishavijay this is design feedback I gave in the call. I’ll also take over to do a last review pass over it after the changes are done. :)

redblom commented:
Screenshots belonging to https://github.com/nextcloud/contacts/pull/4417/commits/66bca00d00442271865ed60da2260dffa6ffaed1


<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-30-26" src="https://github.com/user-attachments/assets/3b0ebd02-0078-412c-8fbb-315eec0dda2f" />
OCM invitations menu moved to top.


<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-32-09" src="https://github.com/user-attachments/assets/88f48455-b120-4e56-babb-a75805f9a4cc" />
New invitation form: label field removed, title fixed, removed extra container


<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-32-39" src="https://github.com/user-attachments/assets/058258e4-826e-4932-9961-0846dda3627a" />
Accept invite form: title fixed


<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-33-13" src="https://github.com/user-attachments/assets/299323ac-6006-42a8-a754-04e4395be348" />
All invitations moved

<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-33-45" src="https://github.com/user-attachments/assets/203aebaf-44c5-43c3-8bdf-1b1054535c78" />
Options to create/accept invites on empty invitations page

<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-34-55" src="https://github.com/user-attachments/assets/5a8c9770-21d8-4a63-a4d4-b03136594504" />
Correct message when trying to accept an invite you've created yourself


<img width="1859" height="910" alt="Screenshot from 2026-06-26 15-53-52" src="https://github.com/user-attachments/assets/4a7c7142-98b2-449c-ad44-ec59a2ce5613" />
WAYF page: new title, removed sub text

jancborchardt reviewed (commented):
Nice @redblom! Looks very good and basically good to go!

Only have 3 tiny pieces of feedback based on those changes, but they are also not a blocker:
- [x] The nav entry should be called "External invitations" instead of "All invitations" as that is clearer, no?
- [x] In the modal for "Create invitation", the button says "Send invite", but everywhere else we use the language "invitation". So better to use "Send invitation" here too.
- [ ] On the emptycontent page "There are no invites", there are 2 primary buttons. There should only ever be 1 primary button, in this case it makes most sense to be "Create invitation". (And they can be next to each other rather than below each other.)

redblom commented:
@jancborchardt - Thank you for the fast review !!

> Nice @redblom! Looks very good and basically good to go!
> 
> Only have 3 tiny pieces of feedback based on those changes, but they are also not a blocker:
> 
>     * [ ]  The nav entry should be called "External invitations" instead of "All invitations" as that is clearer, no?
> 
>     * [ ]  In the modal for "Create invitation", the button says "Send invite", but everywhere else we use the language "invitation". So better to use "Send invitation" here too.
> 
>     * [ ]  On the emptycontent page "There are no invites", there are 2 primary buttons. There should only ever be 1 primary button, in this case it makes most sense to be "Create invitation". (And they can be next to each other rather than below each other.)

Screenshots with implementation:
<img width="1859" height="910" alt="Screenshot from 2026-06-28 12-05-41" src="https://github.com/user-attachments/assets/e2d977f3-1c2d-4194-ac2a-aa87f09d1a2e" />

<img width="1859" height="910" alt="Screenshot from 2026-06-28 12-06-33" src="https://github.com/user-attachments/assets/1fe413fd-c097-4ce3-9982-51c10dfaf40f" />

miaulalala commented:
@jancborchardt can you +1 if you're happy with the changes? @hamza221 I'd apprechiate a review as well. Should be merged only after the server PR is in.

jancborchardt reviewed (approved):
Thought I already approved, here you go. :)

ChristophWurst commented:
> Should be merged only after the server PR is in.

Didn't look at the full diff but sounds like this needs a version guard then. The Contacts app is distributed via the app store. It may be updated earlier than an update updates the server. That must not lead to an incompatibility.

redblom commented:
> > Should be merged only after the server PR is in.
> 
> Didn't look at the full diff but sounds like this needs a version guard then. The Contacts app is distributed via the app store. It may be updated earlier than an update updates the server. That must not lead to an incompatibility.

Firstly, the unit tests that test the OCM invitation flow do not account for nextcloud versions below 35 and I guess that's why those tests currently fail. So that needs to be fixed. 
... assuming we don't want `nextcloud min-version` to be `35`. 

Second, @ChristophWurst, what version guard implementation do you have in mind?

ChristophWurst commented:
A guard to make this app usable when deployed on 33/34. Perhaps the feature can just be hidden for those.

redblom commented:
> A guard to make this app usable when deployed on 33/34. Perhaps the feature can just be hidden for those.

Perfect. That has been accounted for. ~~Will do~~.

The feature is hidden by default.

Also there are two backport PRs for 33/34 ([#62460](https://github.com/nextcloud/server/pull/62460), #[62486](https://github.com/nextcloud/server/pull/62486) ) for this waiting for acceptance.

redblom committed:
feat(OCM-Invites): Implement exchanging cloud IDs through OCM invitation workflow
see https://github.com/cs3org/OCM-API/blob/v1.2.1/IETF-RFC.md

Features:
 - Button to invite remote users to exchange cloudIDs.
 - Button to manually accept invite to exchange cloudIDs.
 - WAYF page allowing the receiver of the invite to select provider and accept the invitation.
 - Listing of open invitations.
 - Option to resend, revoke open invitations.
 - Administer invitation settings

Signed-off-by: Antoon P. <antoon.prins@surf.nl>
Co-authored-by: Micke Nordin <kano@sunet.se>
Co-authored-by: Mahdi Baghbani <mahdi-baghbani@azadehafzar.io>

redblom commented:
Rebase.

redblom commented:
Merging [#62460](https://github.com/nextcloud/server/pull/62460) and #[62486](https://github.com/nextcloud/server/pull/62486) should make the CI build succeed.

redblom commented:
```
Run composer run psalm -- --threads=1 --monochrome --no-progress --output-format=github
> psalm '--threads=1' '--monochrome' '--no-progress' '--output-format=github'
Error: lib/Controller/FederatedInvitesController.php:299:40: TypeDoesNotContainType: '' cannot be identical to non-empty-string (see https://psalm.dev/056)
Error: lib/Controller/FederatedInvitesController.php:299:75: TypeDoesNotContainType: '' cannot be identical to non-empty-string (see https://psalm.dev/056)
------------------------------
2 errors found
```
This does not seem to occur locally. What's going on here ??

ChristophWurst commented:
> This does not seem to occur locally. What's going on here ??

CI installs stub packages dynamically. You might have Psalm running against another version. See here https://github.com/nextcloud/contacts/blob/59ea0379f2b9efdbb73fa7362f02af4723cd3636/.github/workflows/psalm-matrix.yml#L73-L74

hamza221 commented on lib/Service/SocialApiService.php:
```diff
@@ -107,30 +111,22 @@ protected function addPhoto(array &$contact, string $imageType, string $photo) {
 				$addressBook = $ab;
 			}
 		}
-
-		$addressBookIsUpdatable = $addressBook !== null
-			&& ($addressBook->getPermissions() & Constants::PERMISSION_UPDATE);
-
-		if (!$addressBookIsUpdatable) {
-			return null;
-		}
-
```
Wouldn't this cause `updateContact`down below to fail on readonly addressbooks ?

redblom replied:
@hamza221 - I'm on all of these. There's definitely some technical dept that has developed.

redblom replied:
Done. Reverted.

hamza221 commented on lib/Service/SocialApiService.php:
```diff
@@ -252,7 +252,131 @@ public function updateContact(string $addressBookId, string $contactId, ?string
+	 *                    that cloud id already exists or there is no suitable
+	 *                    writable address book for the inviter
+	 */
+	public function createFederatedContact(string $cloudId, string $email, string $name, string $userId): ?array {
```
is this used anywhere ?

redblom replied:
Removed, fixed tests.

hamza221 commented on src/components/AppNavigation/Settings/SettingsImportContacts.vue:
```diff
@@ -307,6 +316,10 @@ export default {
+
+		getOcmInvitesActiveClass() {
+
+		},
```
Leftover?

redblom replied:
Removed

hamza221 commented on src/components/Ocm/OcmInviteAccept.vue:
```diff
@@ -0,0 +1,102 @@
+<!--
+  - SPDX-FileCopyrightText: 2018 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
  - SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
```

redblom replied:
All fixed.

hamza221 commented on src/components/Ocm/OcmInviteDetails.vue:
```diff
@@ -0,0 +1,443 @@
+<!--
+  - SPDX-FileCopyrightText: 2018 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
  - SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
```

redblom replied:
All fixed.
